### PR TITLE
CLOUDP-285964: Adds IPA rule 104 - Resource has get

### PR DESCRIPTION
**Pull request by lovisaberggren**
## Proposed changes

- IPA 104 resource must have GET rule implementation
- Rule test with Jest `npm run test`
- Run IPA validation with `npm run ipa-validation`

Example output:

```
warning  xgen-IPA-104-resource-has-GET  APIs must provide a get method for resources. http://go/ipa/117  paths./api/atlas/v2/orgs/{orgId}/users/{userId}/roles
```

_Jira ticket:_ [CLOUDP-285964](https://jira.mongodb.org/browse/CLOUDP-285964)

### Follow up

- Run tests and IPA validation on PRs
- Exceptions

### CODE REVIEW

File: tools/spectral/ipa/rulesets/functions/utils/resourceEvaluation.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+export function isChild(path) {
+  return path.endsWith('}');
+}
+
+export function isCustomMethod(path) {
+  return path.includes(':');
+}
+
+/**
+ * Checks if a resource is a singleton resource based on the paths for the
```
Comment (yelizhenden-mdb) [inline tools/spectral/ipa/rulesets/functions/utils/resourceEvaluation.js:10]: Nit: Could we add a comment explaining what 'singleton resource' and 'standard resource' mean for clarity?

File: tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          post: {},
+        },
+        '/singleton': {
+          patch: {},
+        },
+      },
+    },
+    errors: [
+      {
+        code: 'xgen-IPA-104-resource-has-GET',
```
Comment (yelizhenden-mdb) [inline tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js:67]: Q: Why does `/custom` cause an error if we are not checking `child` and `custom` methods? Same question for `/standard` as well

Comment (yelizhenden-mdb) [reply tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js:67]: Is it because we’re leaving all the checks to the parent, which is responsible for handling both child routes and custom methods?

File: tools/spectral/ipa/__tests__/__helpers__/testRule.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import * as fs from 'node:fs';
+import * as path from 'node:path';
+import { describe, expect, it } from '@jest/globals';
+import { Spectral, Document } from '@stoplight/spectral-core';
+import { httpAndFileResolver } from '@stoplight/spectral-ref-resolver';
+import { bundleAndLoadRuleset } from '@stoplight/spectral-ruleset-bundler/with-loader';
+
+const rulesetPath = path.join(__dirname, '../..', 'ipa-spectral.yaml');
+
+export default (ruleName, tests) => {
```
Comment (yelizhenden-mdb) [inline tools/spectral/ipa/__tests__/__helpers__/testRule.js:10]: Nit: Should we add this repo structure for unit tests to the spec?

File: tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js (hunk 2 above)
Comment (lovisaberggren) [reply tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js:67]: Yes, each "parent" path is considered the id for a resource, and the rule checks that the resource has a get (in children or if it's a singleton -> on itself). We don't require custom methods to have a GET, they can be either GET or POST (https://docs.devprod.prod.corp.mongodb.com/ipa/109), but in the case here `custom/{exampleId}` should have a GET, so we get an error for `/custom` path since it's the resource parent. Let me know if that makes sense!

File: tools/spectral/ipa/__tests__/__helpers__/testRule.js (hunk 3 above)
Comment (lovisaberggren) [reply tools/spectral/ipa/__tests__/__helpers__/testRule.js:10]: I can add it to the appendix 👍

Comment (lovisaberggren) [reply tools/spectral/ipa/__tests__/__helpers__/testRule.js:10]: https://docs.google.com/document/d/1hcYKgnD7FUGII-3BP4TgiBc9rDcNyd2LMuzbxqFUl1Q/edit?tab=t.0#bookmark=id.irlwpe8w17gn

File: tools/spectral/ipa/rulesets/functions/eachResourceHasGetMethod.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    }
+  } else if (isStandardResource(resourcePaths)) {
+    if (!hasGetMethod(oas.paths[resourcePaths[1]])) {
+      return [
+        {
+          message: ERROR_MESSAGE,
+        },
+      ];
+    }
+  }
```
Comment (yelizhenden-mdb) [inline tools/spectral/ipa/rulesets/functions/eachResourceHasGetMethod.js:36]: Q: What if it is not a singleton and standard resource?(The equivalent of `/custom` in the tests). Don't we check for `get` method at all?

Comment (lovisaberggren) [reply tools/spectral/ipa/rulesets/functions/eachResourceHasGetMethod.js:36]: Currently we have no resources that fall into neither standard nor singleton (with the checks currently implemented)

File: tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
+import testRule from './__helpers__/testRule';
+import { DiagnosticSeverity } from '@stoplight/types';
+
+testRule('xgen-IPA-104-resource-has-GET', [
```
Comment (gssbzn) [inline tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js:4]: can you add a case with nested resource?
`/groups/{groupId}/clusters`? and a singleton like `/groups/{groupId}/settings`?

Comment (lovisaberggren) [reply tools/spectral/ipa/__tests__/eachResourceHasGetMethod.test.js:4]: Done!

File: tools/spectral/ipa/rulesets/functions/utils/resourceEvaluation.js
Comment (gssbzn) [inline tools/spectral/ipa/rulesets/functions/utils/resourceEvaluation.js:1]: can we unit test these, and if not can we add some examples to things like `isSingletonResource` or `isStandardResource` it's a bit hard to follow the intention of the method and some examples either in text or as actual tests would be of great help in the future if we need to debug or change those functions

Comment (lovisaberggren) [reply tools/spectral/ipa/rulesets/functions/utils/resourceEvaluation.js:1]: Sure, makes sense, I'll add unit tests for them as well (getResourcePaths, isStandardResource and isSingletonResource as they are probably the tricky ones)

File: tools/spectral/ipa/rulesets/functions/utils/resourceEvaluation.js [hunk 6]
```diff
@@ -58,6 +58,9 @@ export function hasGetMethod(pathObject) {
  */
 export function getResourcePaths(parent, allPaths) {
   const childPathPattern = new RegExp(`^${parent}/{[a-zA-Z]+}$`);
-  const customMethodPattern = new RegExp(`^${parent}/{[a-zA-Z]+}:+[a-zA-Z]+$`);
-  return allPaths.filter((p) => parent === p || childPathPattern.test(p) || customMethodPattern.test(p));
+  const customChildMethodPattern = new RegExp(`^${parent}/{[a-zA-Z]+}:+[a-zA-Z]+$`);
+  const customMethodPattern = new RegExp(`^${parent}:+[a-zA-Z]+$`);
+  return allPaths.filter(
+    (p) => parent === p || childPathPattern.test(p) || customMethodPattern.test(p) || customChildMethodPattern.test(p)
```
Comment (lovisaberggren) [inline tools/spectral/ipa/rulesets/functions/utils/resourceEvaluation.js:64]: The added tests revealed that we were missing a custom case, as it can be either `/path/{id}:method` or `/path:method`

**Review by gssbzn — APPROVED** [review]
LGTM

**Review by yelizhenden-mdb — APPROVED** [review]
LGTM, exciting! 🚀
